File: CS/queue/ruby/queue.rb
require_relative 'node'

class Queue
    attr_reader :size

    def initialize(max_size = nil)
        @head = nil
        @tail = nil
        @size = 0
        @max_size = max_size
    end

    def enqueue(value)
        raise 'No more room.' unless has_space?

        new_node = Node.new(value)

        if empty?
            @head = @tail = new_node
        else
            @tail.next_node = new_node
            @tail = new_node
        end

        @size += 1
    end

    def dequeue
        raise 'Queue is empty.' if empty?

        node_to_remove = @head
        @head = @head.next_node

        @tail = nil if @head.nil?
        @size -= 1

        node_to_remove.value
    end

    def peek
        empty? ? nil : @head.value
    end

    def has_space?
        @max_size.nil? || @max_size > @size
    end

    def empty?
        @size == 0
    end
end

File: CS/queue/python/queue.py
from node import Node

class Queue:
    
    def __init__(self, max_size=None):
        self.head = None
        self.tail = None
        self.size = 0
        self.max_size = max_size
    
    def enqueue(self, value):
        if not self.has_space(): raise Exception('No more room.')
        
        new_node = Node(value)
        
        if self.is_empty():
            self.head = self.tail = new_node
        else:
            self.tail.set_next_node(new_node)
            self.tail = new_node
        
        self.size += 1
    
    def dequeue(self):
        if self.is_empty(): raise Exception('Queue is empty.')
        
        node_to_remove = self.head
        
        if self.size == 1: self.head = self.tail = None
        else: self.head = node_to_remove.get_next_node()
        
        self.size -= 1
        return node_to_remove.get_value()
    
    def peek(self):
        if self.is_empty(): return None
        
        return self.head.get_value()
    
    def get_size(self):
        return self.size
    
    def has_space(self):
        return not self.max_size or self.max_size > self.size

    def is_empty(self):
        return self.size == 0